File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    
    public int f(int i , int buy , int[] p , int dp[][]) {
        int n = p.length ;
        if(i >= n ) return 0 ;
        int ans = 0 ;
        if(dp[i][buy] != -1) return dp[i][buy];
        if(buy == 1){
            int x = f(i+1 , 0 , p , dp) - p[i];
            int y = f(i+1 , 1 , p , dp); 
            ans = Math.max(x , y) ;
        }
        else if(buy ==0 ){
            int y = f(i+1 , 0 , p , dp) ;
            int x = f(i+2 , 1 , p , dp) + p[i];
            ans = Math.max(x , y );
        }
        return dp[i][buy] =  ans ; 
    }
    public int maxProfit(int[] prices) {
        int dp[][] = new int[prices.length][2];
        
        for(int i = 0 ; i < prices.length ; i++){
            dp[i][0] = -1 ; 
            dp[i][1] = -1 ;
        }
        return f(0 , 1 , prices  , dp ) ;
        
    }
}

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int helper(vector<int> &a , int i , int buy , vector<vector<int>> &dp) {
        if(i >= a.size()) return 0 ;
        if(dp[i][buy] != -1) return dp[i][buy];
        
        int ans = 0 ; 
        
        if (buy == 1) {
            int x = (-1 * a[i] )+ helper(a , i + 1 , 0 , dp);
            int y = 0 + helper(a , i + 1 , 1 ,dp);
            ans = max(x , y);
        }
        
        else if (buy == 0) {
            int x = a[i] + helper(a , i+2 , 1 ,dp);
            int y = 0 + helper(a , i+1 , 0 ,dp);
            ans = max(x , y);
        }

        return dp[i][buy] = ans ; 
    }
    int maxProfit(vector<int>& a) {
        
        int n = a.size() ; 
        vector<vector<int>> dp(n+1 , vector<int>(2 , -1));
        return helper(a, 0 , 1 , dp) ;
    }
};